### PR TITLE
[Basic] Add pretty printing in JSON

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@ddunbar Do you think we need a toggle for pretty printing or just doing this always is OK?

**Comment by ddunbar** [comment]
I think we need a toggle, since in some cases the extra size of pretty printing is undesirable.

File: Sources/Basic/JSON.swift [hunk 1]
```diff
@@ -122,23 +126,22 @@ extension JSON: ByteStreamable {
...
+            stream <<< "[" <<< "\n"
             for (i, item) in contents.enumerated() {
-                if i != 0 { stream <<< ", " }
-                stream <<< item
+                if i != 0 { stream <<< "," <<< "\n" }
+                stream <<< " ".repeating(n: indent + 2)
+                item.write(to: stream, indent: indent + 2)
             }
-            stream <<< "]"
+            stream <<< "\n" <<< " ".repeating(n: indent) <<< "]"
```
Comment (ddunbar) [inline Sources/Basic/JSON.swift]: We should get an OutputByteStream way to write a repeating sequence -- this way is inefficient as it creates the string just to throw it away.

The simple way to do this is to add a `Format` style wrapper type for writing a repeating character.

**Comment by aciidb0mb3r** [comment]
updated

**Comment by ddunbar** [comment]
Did you check the performance impact of this? I suspect that the way it is implemented will cause a substantial regression in JSON output performance...

**Comment by aciidb0mb3r** [comment]
Just measured it, don't see any regression

File: Sources/Basic/JSON.swift [hunk 2]
```diff
@@ -122,23 +130,22 @@ extension JSON: ByteStreamable {
...
+            stream <<< "[" <<< (shouldIndent ? "\n" : "")
             for (i, item) in contents.enumerated() {
-                if i != 0 { stream <<< ", " }
-                stream <<< item
+                if i != 0 { stream <<< "," <<< (shouldIndent ? "\n" : " ") }
+                stream <<< indentStreamable(offset: 2)
+                item.write(to: stream, indent: indent.flatMap {$0 + 2})
             }
-            stream <<< "]"
+            stream <<< (shouldIndent ? "\n" : "") <<< indentStreamable() <<< "]"
```
Comment (aciidb0mb3r) [inline Sources/Basic/JSON.swift]: it indents the value passed in param. And offset is used to indent more than the passed value (indent+offset)

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Pull request merged**